src/stockpile.clj

```diff
@@ -0,0 +1,364 @@
+
+(defn- create-tmp-file [parent]
+  (Files/createTempFile (as-path parent) "tmp-" ""
+                        (into-array FileAttribute [])))
+
+(def ^:private osx? (.contains (System/getProperty "os.name") "OS X"))
+
+(when osx?
+  (binding [*out* *err*]
+    (println (str "Warning: stockpile may not be durable on OS X:"
+                  " https://bugs.openjdk.java.net/browse/JDK-8080589"))))
+
+(defn fsync [x metadata?]
+  (with-open [fc (FileChannel/open (as-path x)
+                                   (into-array OpenOption []))]
+    (.force fc metadata?)))
+
+(defn- atomic-move [src dest]
+  (Files/move (as-path src) (as-path dest)
+              (into-array [StandardCopyOption/ATOMIC_MOVE])))
+
+(defn- rename-durably [src dest sync-parent?]
+  "If possible, atomically renames src to dest (each of which may be a
+  File, Path, or String).  If dest already exists, on some platforms
+  the replacement will succeed, and on others it will throw an
+  IOException.  The rename may also fail with
+  AtomicMoveNotSupportedException (perhaps if src and dest are on
+  different filesystems).  See java.nio.file.Files/move for additional
+  information.  Fsyncs the dest parent directory to make the final
+  rename durable unless sync-parent? is false (presumably the caller
+  will ensure the sync)."
+  (atomic-move src dest)
+  (when sync-parent?
+    (fsync (.getParent (as-path dest)) true)))
+
+(defn- durably-establish [path write-content sync-parent?]
+  "Calls (write-content temp-path) and durably stores the resulting
+  temp-file contents at path (a File, Path, or String).  Fsyncs the
+  parent directory of dest to make the final rename durable unless
+  sync-parent? is false (presumably the caller will ensure the sync).
+  See rename-durably for additional information.  As compared
+  to (store q stream ...), the write function here allows the caller
+  more control over what happens when something goes wrong (say they
+  know how they might free up space in the filesystem if write fails)."
+  (let [parent (.getParent (as-path path))
+        tmp (create-tmp-file parent)]
+    (try
+      (write-content tmp)
+      (fsync tmp false)
+      (rename-durably tmp path sync-parent?)
+      (catch Exception ex
+        (Files/deleteIfExists tmp)
+        (throw ex)))))
+
+(defn- filename->entry [name decode-meta]
+  "Returns an entry if name can be parsed as such, i.e. either as
+  an integer or integer-metadata, nil otherwise."
+   (let [dash (.indexOf name "-")]
+    (if (= -1 dash)
+      (parse-integer name)
+      ;; Perhaps it has metadata
+      (when-let [id (parse-integer (subs name 0 dash))]
+        (when-let [meta (try
+                          (decode-meta (subs name (inc dash)))
+                          (catch IllegalArgumentException ex
+                            nil))]
+          (->MetaEntry id meta))))))
+
+(defn- existing-entries [q decode-meta]
+  (let [dirstream (Files/newDirectoryStream (.resolve (:directory q) "q"))]
```

Comment:
It seems like this would be a useful function as we do this many times in the code